virttest/qemu_vm.py

```diff
@@ -4266,7 +4266,20 @@ def _shell_reboot(session, timeout):
             raise virt_vm.VMRebootError("Unknown reboot method: %s" % method)
         if _check_go_down is None:
             if isinstance(self.monitor, qemu_monitor.QMPMonitor):
-                _check_go_down = partial(self.monitor.get_event, "RESET")
+                original_monitor = self.monitor
+
+                def _check_go_down():
+                    if self.monitor != original_monitor:
+                        logging.warning("Monitor changed during check for "
+                                        "reboot, checking for reboot is not "
+                                        "reliable. Assuming it already "
+                                        "rebooted.")
+                        return True
```

Comment:
Here we have two choices:

1. always return `True` by monitor change, then getting false-positive results is the side effect.
2. continue looking up the event from destination side, then getting false-negative results is the side effect (in this case, the event has been sent to source side but we missed that).

I prefer the second one because it would less likely give the wrong results than the first one.

Reply:
Actually I was trying it with a VM that took long to reboot and almost always it rebooted in the wrong time. That is why I chose to return `True`. Note that we also return True on human monitor by default so it's nothing uncommon.

Anyway even if we get false-positive, VM should not accept new connections while rebooting, therefor we should get failed-to-boot. I mean currently about 10% of my jobs are failing because of this, changing it to default False it would be something between 1-10% so it should be better but honestly I'd prefer wrong failed message and ~0% of failures than 1-10% of failures where I'd have to look whether it was because of missed qmp event or the machine actually refusing to go down.

So the question is, do you just slightly prefer it, or is it a strong nack? It should not introduce false-positives, it could just occasionally report failed-to-boot instead of failed-to-shutdown-before-boot.

Reply:
Sorry for the delayed response.

> Anyway even if we get false-positive, VM should not accept new connections while rebooting, therefor we should get failed-to-boot.

Well, "failed-to-boot" (I am not sure about what you mean by this error, so I understand it as a kind of `Exception`) is just one of the moderate errors, I mean at least the function would raise exceptions in that case. The worst thing is the function would successfully return with a new session while the guest is just at the beginning to commit a reboot, if the sequence goes like as the follow it could happen:

```
 time \ routines |           vm.reboot()          | migration |        os system service        |
-----------------+--------------------------------+-----------+---------------------------------+
        t1       |   _reboot() (shell) returned   |           |                                 |
        t2       |                                | completed | preparing to shut down services |
        t3       | _check_go_down() returned True |           |                                 |
        t4       |   calling vm.wait_for_login()  |           |                                 |
        t5       |        reboot() returned       |           |        shutting down sshd       |
        t6       |                                |           |          shut down sshd         |
        t7       |                                |           |            rebooting            |
```

And I guess we may meet this more in case of rdma migration.

> I mean currently about 10% of my jobs are failing because of this, changing it to default False it would be something between 1-10% so it should be better but honestly I'd prefer wrong failed message and ~0% of failures than 1-10% of failures where I'd have to look whether it was because of missed qmp event or the machine actually refusing to go down.
> 
> So the question is, do you just slightly prefer it, or is it a strong nack? It should not introduce false-positives, it could just occasionally report failed-to-boot instead of failed-to-shutdown-before-boot.


Actually, I do want to request the changes, but I'd like to know your opinions at first.

Thanks.

Reply:
I'd call that unlikely, but I'm not strongly tied to this solution and I understand your concern. Let me simply keep using the current monitor, which means we can miss the event, but we should not declare the reboot complete unless the RESET event is emitted.